src/Stratis.Bitcoin.Features.Consensus/CoinViews/DBreezeCoinView.cs

```diff
@@ -319,49 +322,92 @@ public Task<RewindData> GetRewindData(int height)
         }
 
         /// <inheritdoc />
-        public Task<uint256> RewindAsync()
+        public Task<uint256> RewindAsync(int? targetHeight = null)
         {
             Task<uint256> task = Task.Run(() =>
             {
-                uint256 res = null;
+                int currentHeight;
+
                 using (DBreeze.Transactions.Transaction transaction = this.CreateTransaction())
                 {
-                    transaction.SynchronizeTables("BlockHash", "Coins", "Rewind");
-                    if (this.GetRewindIndex(transaction) == 0)
+                    currentHeight = this.GetRewindIndex(transaction);
+                    targetHeight = targetHeight ?? (currentHeight - 1);
+
+                    if (targetHeight <= 0)
                     {
+                        transaction.SynchronizeTables("BlockHash", "Coins", "Rewind");
                         transaction.RemoveAllKeys("Coins", true);
+                        transaction.RemoveAllKeys("Rewind", true);
                         this.SetBlockHash(transaction, this.network.GenesisHash);
 
-                        res = this.network.GenesisHash;
+                        return this.network.GenesisHash;
                     }
-                    else
+                }
+
+                RewindData firstRewindData = null;
+
+                while (targetHeight < currentHeight)
+                {
+                    // Do a batch of blocks.
+                    var keysToRemove = new List<int>();
+                    var transactionsToRemove = new List<uint256>();
+                    var outputsToRestore = new List<UnspentOutputs>();
+
+                    using (DBreeze.Transactions.Transaction transaction = this.CreateTransaction())
                     {
                         transaction.ValuesLazyLoadingIsOn = false;
+                        int firstBlockToRewind = Math.Max(currentHeight - MaxBlocksToRewind + 1, (int)targetHeight + 1);
 
-                        Row<int, byte[]> firstRow = transaction.SelectBackward<int, byte[]>("Rewind").FirstOrDefault();
-                        transaction.RemoveKey("Rewind", firstRow.Key);
-                        var rewindData = this.dBreezeSerializer.Deserialize<RewindData>(firstRow.Value);
-                        this.SetBlockHash(transaction, rewindData.PreviousBlockHash);
-
-                        foreach (uint256 txId in rewindData.TransactionsToRemove)
+                        for (int height = firstBlockToRewind; height <= currentHeight; height++)
                         {
-                            this.logger.LogTrace("Outputs of transaction ID '{0}' will be removed.", txId);
-                            transaction.RemoveKey("Coins", txId.ToBytes(false));
+                            Row<int, byte[]> row = transaction.Select<int, byte[]>("Rewind", height);
+                            if (!row.Exists)
+                                continue;
```

Comment:
This should probably throw It can not be we will not have rewind data IMO.

Reply:
Refactored.